oauth2/src/main/scala/tsec.oauth2/provider/AuthorizationRequest.scala

```diff
@@ -0,0 +1,54 @@
+package tsec.oauth2.provider
+
+import cats.implicits._
+import java.util.Base64
+
+import scala.collection.immutable.TreeMap
+import scala.util.Try
+
+case class ClientCredential(clientId: String, clientSecret: Option[String])
+
+case class AuthorizationRequest(headers: Map[String, Seq[String]], params: Map[String, Seq[String]]){
+  private val orderedHeaders = new TreeMap[String, Seq[String]]()(Ordering.by(_.toLowerCase)) ++ headers
+
+  def scope: Option[String] = params.get(SCOPE).flatMap(_.headOption)
+
+  def grantHandler(isClientCredRequiredForPasswordGrantType: Boolean): Either[OAuthError, GrantHandler] = for{
+    header <- params.get(GRANT_TYPE).flatMap(_.headOption).toRight(InvalidRequest("Missing grant type"))
+    modified = if(header.toLowerCase == PASSWORD) PASSWORD+isClientCredRequiredForPasswordGrantType else header.toLowerCase
+    res <- GrantHandler.strToGrantHandler.get(modified).toRight(UnsupportedGrantType(s"unsupported grant type: $header"))
+  }yield res
+
+  def parseClientCredential: Either[OAuthError, ClientCredential] = {
+    val authHeader = for{
+      h <- orderedHeaders.get("authorization").flatMap(_.headOption).toRight[OAuthError](InvalidClient("Missing authorization header"))
+      matcher <- """^\s*Basic\s+(.+?)\s*$""".r.findFirstMatchIn(h).toRight(InvalidAuthorizationHeader)
+      cred = matcher.group(1)
+      r <- clientCredentialByAuthorization(cred)
+    }yield r
+    authHeader.fold[Either[OAuthError, ClientCredential]](
+      e => {
+        e match {
+          case InvalidAuthorizationHeader => InvalidAuthorizationHeader.asLeft[ClientCredential]
+          case _ =>
+            clientCredentialByParam.toRight(InvalidClient(s"Failed to parse client credential from header (${e.description}) and params"))
+        }
+      },
+      Right(_)
+    )
+  }
+
+  private def clientCredentialByAuthorization(s: String): Either[InvalidClient, ClientCredential] =
+    Try(new String(Base64.getDecoder.decode(s), "UTF-8"))
```

Comment:
2 quips:

- [Tsec-common has a `.base64Bytes` helper, for the record.](https://github.com/jmcardon/tsec/blob/master/common/src/main/scala/tsec/common/package.scala#L54), as well as a utf8 one. This could be done by `s.base64Bytes.toUtf8String`. Granted, this is not safe and in a later PR I need to address this.
- I'm _very_ iffy on `Try` but I might be able to admit this could be a good use case for it.